### PR TITLE
RPC Authorization

### PR DESCRIPTION
**Pull request by kishansagathiya**
This PR adds permission policy for raft consensus and adds
permission policy to config(but hides it from user).
    
It proposes minimum set of RPC methods which we have to allow peers in
order for all cluster features to work as expected. It restricts peers
from accessing methods that are not part of the proposed RPC method set.
    
Proposed authorization would work only for raft consensus.
`authorizeWithPolicy` would need to be edited in order to support
 upcoming consensus mechanisms.

#666

### CODE REVIEW

File: rpc_permissions.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+				"Cluster.ConsensusAddPeer":  true,
+				"Cluster.ConsensusRmPeer":   true,
+				"Cluster.ConsensusLogPin":   true,
+				"Cluster.ConsensusLogUnpin": true,
+			},
+		},
+	}
+}
+
+func getPermissionPolicy(name string) permissionPolicy {
```
Comment (hsanjuan) [inline rpc_permissions.go]: Don't make functions for simple oneliners. It makes a oneliner become 3 lines, and it makes you jump to a different place of the file while reading the code.

Comment (kishansagathiya) [reply rpc_permissions.go]: https://github.com/ipfs/ipfs-cluster/pull/710/commits/1cc32c84608dd5f1ddbf2c17191d5e4fad04f52c

File: rpc_permissions.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		if policy == nil || policyName != DefaultPermissionPolicy {
+			return false
+		}
+
+		return policy[all][svc+"."+method]
+	}
+}
+
+func getPermissionMap() map[string]permissionPolicy {
+	return map[string]permissionPolicy{
```
[the remark this answers is not in the document]
Comment (kishansagathiya) [reply rpc_permissions.go]: https://github.com/ipfs/ipfs-cluster/pull/710/commits/1cc32c84608dd5f1ddbf2c17191d5e4fad04f52c

File: rpc_permissions.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+type peerKind int
+
+const (
+	all peerKind = iota
+	trusted
+)
+
+type permissionPolicy map[peerKind]map[string]bool
+
+func (c *Cluster) authorizeWithPolicy() func(pid peer.ID, svc string, method string) bool {
```
[the remark this answers is not in the document]
Comment (kishansagathiya) [reply rpc_permissions.go]: https://github.com/ipfs/ipfs-cluster/pull/710/commits/1cc32c84608dd5f1ddbf2c17191d5e4fad04f52c

File: rpc_permissions.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	raft       string = "raft"
+	crdtStrict string = "crdt_strict"
+	crdtSoft   string = "crdt_soft"
+)
+
+type permissionPolicy map[peerKind]map[string]bool
+type authorizer struct {
+	policyName string
+	policy     permissionPolicy
+	trusted    []peer.ID
```
Comment (kishansagathiya) [inline rpc_permissions.go]: This should be a map.

File: rpc_permissions.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+			return policy[all][svc+"."+method]
+		}
+
+		return false
+	}
+}
+
+func (a *authorizer) addPeers(peer peer.ID) {
+	a.trusted = append(a.trusted, peer)
+}
```
Comment (kishansagathiya) [inline rpc_permissions.go]: Probably be able to add multiple peers!

**Comment by kishansagathiya** [comment]
Can I get a quick review for this?

**Comment by hsanjuan** [comment]
Thanks @kishansagathiya , I will finish this PR

**Comment by hsanjuan** [comment]
Superseeded at #775

**Pull request closed without merging**